common/expirecontainer.cpp

```diff
@@ -38,24 +40,31 @@ auto ExpireContainerBase::find(const Item& key_item) -> iterator {
 }
 
 void ExpireContainerBase::clear() {
-    for (auto x : ({
-             SCOPED_LOCK(_lock);
-             _list.node = nullptr;
-             Set(std::move(_set));
-         })) {
-        delete x;
+    Set tmp;
+    {
+        SCOPED_LOCK(_lock);
+        _list.node = nullptr;
+        tmp = std::move(_set);
     }
+    for (auto x : tmp)
+        delete x;
 }
 
 uint64_t ExpireContainerBase::expire() {
-    ({
+    if (_lifespan > UINT64_MAX / 2)
+        return 0;
+    intrusive_list<Item> expired;
+    {
         SCOPED_LOCK(_lock);
-        _list.split_by_predicate([&](Item* x) {
+        auto tmp = _list.split_by_predicate([&](Item* x) {
```

Comment:
`_list.split_by_predicate(...)` will produce a new `intrusive_list` object, but `intrusive_list` has no move constructor or move assignment. That makes both `tmp` and pr-value of method result point to same list.

That makes pr-value `_list.split_by_predicate(...)` not empty after assignment, and then as a temporary object, goes to destruct, meets the assertion about intrusive_list `__node == nullptr`.

Reply:
Shall we add move semantics to `intrusive_list`, or just clean its pointer?